### PR TITLE
Stop subprocesses from getting unexpectedly killed

### PR DESCRIPTION
**Pull request by corhere**
The behaviour of `(syscall.SysProcAttr).Pdeathsig` on Linux has some surprising behaviour, as detailed in https://github.com/golang/go/issues/27505. The parent death signal is sent to the child process when the creating _thread_ dies, which in a Go program is going to be an arbitrary thread taken from the goroutine thread pool and recycled for other goroutines to execute on. A goroutine can cause the thread it is running on to exit by wiring itself to the thread with `runtime.LockOSThread()` and returning. Letting the creating thread become available for other goroutines risks having that thread terminated arbitrarily, and the child process with it.

While the problem can in theory be worked around by ensuring that no goroutine returns while wired to a thread, this is not practical. Unpaired `runtime.LockOSThread()` calls could be tucked away in third-party modules, and mistakes happen in first-party code. And there are legitimate use-cases for terminating threads early, which [heavily overlap with the use-cases for `pkg/reexec`](https://www.weave.works/blog/linux-namespaces-golang-followup), which makes it impractical to ban. The only viable solution is to arrange for the threads which are parents of subprocesses started with `Pdeathsig` to not be eligible for other goroutines to execute on them while the subprocess is alive.

**- What I did**
* Audited all calls to `runtime.LockOSThread()` in non-vendor code in the daemon process to ensure they had matching `runtime.UnlockOSThread()` calls. Unmatched calls in reexec functions were left alone.
* Ensured that the threads which create subprocesses with `Pdeathsig` could not be terminated before the subprocess or the daemon process exited, whichever comes first.

**- How I did it**
* `runtime.LockOSThread()` is both the problem and solution

**- How to verify it**
* CI 🟢 

**- Description for the changelog**

* Resolved an issue which could potentially cause daemon-managed containerd or certain utility subprocesses from being prematurely killed on Linux

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by neersighted** [comment]
Immediate thoughts -- this invariant/explanation needs to be in the doc-comment of `reexec.Command()` in `command_linux` of `pkg/reexec`.

**Comment by cpuguy83** [comment]
Oh wow... great find...

I don't like relying on thread locking to handle this.
The reason we set Pdeathsig is to make sure these subprocesses are cleaned up on exit.

Perhaps the way to go here is to have the systemd unit kill all processes on exit... we wouldn't be able to support live-restore when using managed containerd, but that may be a fare trade-off (or maybe we could move containerd to a separate cgroup in that case...).

**Comment by corhere** [comment]
> I don't like relying on thread locking to handle this. The reason we set Pdeathsig is to make sure these subprocesses are cleaned up on exit.

Right, the subprocesses are cleaned up on daemon exit. Thread locking does not break that intended behaviour; it only prevents subprocesses from being spuriously killed _before_ the daemon exits. [No threads outlive a daemon exit](https://github.com/golang/go/blob/dbf174d4b9dd96b225a1f10089478f8257a54f80/src/runtime/sys_linux_amd64.s#L56-L60) or [SIGKILL](https://security.stackexchange.com/a/173036) or [OOM kill](https://www.kernel.org/doc/gorman/html/understand/understand016.html#:~:text=Once%20a%20task%20is%20selected%2C%20the%20list%20is%20walked%20again%20and%20each%20process%20that%20shares%20the%20same%20mm_struct%20as%20the%20selected%20process%20(i.e.%20they%20are%20threads)%20is%20sent%20a%20signal.). `runtime.LockOSThread()` locks the _calling goroutine_ to the thread, nothing more. The thread itself is not "locked". It does not prevent the thread from being terminated in the way you're thinking. It just prevents other goroutines from being executed on the thread, as other goroutines could make the thread terminate early.
```go
func CausePdeathsigChaos() {
        for i := 0; i < 100; i++ {
                go func() {
                        runtime.LockOSThread()
                        return // Runtime terminates the thread currently executing this goroutine.
                }()
        }
}
```

**Review by neersighted — APPROVED** [review]
LGTM (including a manual test with @corhere's Russian roulette code -- I wonder if there's a way we could turn that into an integration test?)

**Comment by corhere** [comment]
I don't think we'll need an explicit integration test as #44210 will bring the chaos all on its own. The existing integration tests blew up spectacularly on that PR before I rebased it on top of this change.

**Comment by corhere** [comment]
@thaJeztah could you be thinking of this one? https://github.com/moby/moby/pull/44210#discussion_r983913235

**Comment by thaJeztah** [comment]
> @thaJeztah could you be thinking of this one? [#44210 (comment)](https://github.com/moby/moby/pull/44210#discussion_r983913235)

Hm.. no, I think there was one where, hm... what was it? Where go funcs would still be created? Perhaps @cpuguy83 recalls 🤔

File: pkg/chrootarchive/archive_unix.go [hunk 1]
```diff
@@ -188,6 +194,12 @@ func invokePack(srcPath string, options *archive.TarOptions, root string) (io.Re
 		return nil, errors.Wrap(err, "error getting options pipe for tar process")
 	}
 
+	// reexec.Command() sets cmd.SysProcAttr.Pdeathsig on Linux, which
+	// causes the started process to be signaled when the creating OS thread
+	// dies. Ensure that the reexec is not prematurely signaled. See
+	// https://go.dev/issue/27505 for more information.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
```
Comment (cpuguy83) [inline pkg/chrootarchive/archive_unix.go:202]: The thread locking here (and cmd.Start/cmd.Wait) needs to be done in a separate goroutine.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: pkg/chrootarchive/archive_unix.go (hunk 1 above)
Comment (corhere) [reply pkg/chrootarchive/archive_unix.go:202]: Aha! That's the one! I'll fix that right up.

Comment (thaJeztah) [reply pkg/chrootarchive/archive_unix.go]: Ah, thanks! Yes that looks like the one I had in mind ❤️

**Review by cpuguy83 — APPROVED** [review]

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**